src/features/soft_navigations/aggregate/interaction.js

```diff
@@ -0,0 +1,159 @@
+import { getInfo } from '../../../common/config/config'
+import { globalScope, initialLocation } from '../../../common/constants/runtime'
+import { generateUuid } from '../../../common/ids/unique-id'
+import { addCustomAttributes, getAddStringContext, nullable, numeric } from '../../../common/serialize/bel-serializer'
+import { cleanURL } from '../../../common/url/clean-url'
+import { NODE_TYPE, INTERACTION_STATUS, INTERACTION_TYPE, API_TRIGGER_NAME } from '../constants'
+import { BelNode } from './bel-node'
+
+/**
+ * link https://github.com/newrelic/nr-querypack/blob/main/schemas/bel/7.qpschema
+ **/
+export class Interaction extends BelNode {
+  id = generateUuid() // unique id that is serialized and used to link interactions with errors
+  initialPageURL = initialLocation
+  oldURL = '' + globalScope?.location
+  newURL = '' + globalScope?.location
+  customName
+  customAttributes = {}
+  customDataByApi = {}
+  queueTime // only used by initialPageLoad interactions
+  appTime // only used by initialPageLoad interactions
+  newRoute
+  /** Internal state of this interaction: in-progress, finished, or cancelled. */
+  status = INTERACTION_STATUS.IP
+  domTimestamp = 0
+  historyTimestamp = 0
+  createdByApi = false
+  keepOpenUntilEndApi = false
+  onDone = []
+  cancellationTimer
+
+  constructor (agentIdentifier, uiEvent, uiEventTimestamp, currentRouteKnown) {
+    super(agentIdentifier)
+    this.belType = NODE_TYPE.INTERACTION
+    this.trigger = uiEvent
+    this.start = uiEventTimestamp
+    this.oldRoute = currentRouteKnown
+    this.eventSubscription = new Map([
+      ['finished', []],
+      ['cancelled', []]
+    ])
+    this.forceSave = this.forceIgnore = false
+    if (this.trigger === API_TRIGGER_NAME) this.createdByApi = true
+  }
+
+  updateDom (timestamp) {
+    this.domTimestamp = (timestamp || performance.now()) // default timestamp should be precise for accurate isActiveDuring calculations
```

Comment:
should performance.now() in these fallbacks use the `now()` implementation in the agent which also utilizes Math.floor() to normalize all the values?

Reply:
So I ran into a problem here that required finer accuracy. The UI event timeStamp can be something like `100.123`. If this event updateDom at `100.456` and that gets rounded down to `100`, then you end up with an interaction with start and end timestamps of `100.123` and `100` (before it's harvested and rounded down in serialization), which causes problems.
Sure, we can round down the event timeStamp too so that it's `100` and `100` during runtime. However, consider an ajax request that then fires at `100.789`. If you round this down and assign it to the interaction, that's incorrect. If you don't round this down, then an ajax request that fires at `100.234` which should be attributed to the interaction cannot seek it.